osu.Server.Spectator/Entities/ConnectionState.cs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.SignalR;
+using osu.Server.Spectator.Extensions;
+
+namespace osu.Server.Spectator.Entities
+{
+    public class ConnectionState
+    {
+        /// <summary>
+        /// The unique ID of the JWT the user is using to authenticate.
+        /// This is used to control user uniqueness.
+        /// </summary>
+        public readonly string TokenId;
+
+        /// <summary>
+        /// The connection IDs of the user.
+        /// </summary>
+        /// <remarks>
+        /// In SignalR, connection IDs are unique per user, <em>and</em> per hub instance.
+        /// Therefore, to keep track of all of them, a dictionary is necessary.
```

Comment:
Here's a diff with this suggestion applied, alongside some other xmldoc and code structure improvements. Please apply as you see fit:

```diff
diff --git a/osu.Server.Spectator/ConcurrentConnectionLimiter.cs b/osu.Server.Spectator/ConcurrentConnectionLimiter.cs
index 9c25535..835143f 100644
--- a/osu.Server.Spectator/ConcurrentConnectionLimiter.cs
+++ b/osu.Server.Spectator/ConcurrentConnectionLimiter.cs
@@ -42,34 +42,41 @@ namespace osu.Server.Spectator
 
                 using (var userState = await connectionStates.GetForUse(userId, true))
                 {
-                    if (context.Context.GetTokenId() == userState.Item?.TokenId)
+                    if (userState.Item == null)
                     {
+                        log(context, "connection from first client instance");
+                        userState.Item = new ConnectionState(context);
+                        return;
+                    }
+
+                    if (context.Context.GetTokenId() == userState.Item.TokenId)
+                    {
+                        // The assumption is that the client has already dropped the old connection,
+                        // so we don't bother to ask for a disconnection.
+
                         log(context, "subsequent connection from same client instance, registering");
+
+                        // Importantly, this will replace the old connection, ensuring it cannot be
+                        // used to communicate on anymore.
                         userState.Item.RegisterConnectionId(context);
                         return;
                     }
 
-                    if (userState.Item != null)
+                    log(context, "connection from new client instance, dropping existing state");
+
+                    foreach (var hubType in stateful_user_hubs)
                     {
-                        log(context, "connection from new client instance, dropping existing state");
+                        var hubContextType = typeof(IHubContext<>).MakeGenericType(hubType);
+                        var hubContext = serviceProvider.GetRequiredService(hubContextType) as IHubContext;
 
-                        foreach (var hubType in stateful_user_hubs)
+                        if (userState.Item.ConnectionIds.TryGetValue(hubType, out var connectionId))
                         {
-                            var hubContextType = typeof(IHubContext<>).MakeGenericType(hubType);
-                            var hubContext = serviceProvider.GetRequiredService(hubContextType) as IHubContext;
-
-                            if (userState.Item.ConnectionIds.TryGetValue(hubType, out var connectionId))
-                            {
-                                hubContext?.Clients.Client(connectionId)
-                                          .SendCoreAsync(nameof(IStatefulUserHubClient.DisconnectRequested), Array.Empty<object>());
-                            }
+                            hubContext?.Clients.Client(connectionId)
+                                      .SendCoreAsync(nameof(IStatefulUserHubClient.DisconnectRequested), Array.Empty<object>());
                         }
-
-                        log(context, "existing state dropped");
                     }
-                    else
-                        log(context, "connection from first client instance");
 
+                    log(context, "existing state dropped");
                     userState.Item = new ConnectionState(context);
                 }
             }
diff --git a/osu.Server.Spectator/Entities/ConnectionState.cs b/osu.Server.Spectator/Entities/ConnectionState.cs
index dff8e00..80615f7 100644
--- a/osu.Server.Spectator/Entities/ConnectionState.cs
+++ b/osu.Server.Spectator/Entities/ConnectionState.cs
@@ -8,6 +8,9 @@ using osu.Server.Spectator.Extensions;
 
 namespace osu.Server.Spectator.Entities
 {
+    /// <summary>
+    /// Maintains the connection state of a single client (notably, client, not user) across multiple hubs.
+    /// </summary>
     public class ConnectionState
     {
         /// <summary>
@@ -17,11 +20,13 @@ namespace osu.Server.Spectator.Entities
         public readonly string TokenId;
 
         /// <summary>
-        /// The connection IDs of the user.
+        /// The connection IDs of the user for each hub type.
         /// </summary>
         /// <remarks>
-        /// In SignalR, connection IDs are unique per user, <em>and</em> per hub instance.
-        /// Therefore, to keep track of all of them, a dictionary is necessary.
+        /// In SignalR, connection IDs are unique per connection.
+        ///
+        /// Because we use multiple hubs and a user is expected to be connected to each hub,
+        /// we use a dictionary to track connections across all hubs for a specific user.
         /// </remarks>
         public readonly Dictionary<Type, string> ConnectionIds = new Dictionary<Type, string>();
 
@@ -32,6 +37,10 @@ namespace osu.Server.Spectator.Entities
             RegisterConnectionId(context);
         }
 
+        /// <summary>
+        /// Registers the provided hub/connection context, replacing any existing connection for the hub type.
+        /// </summary>
+        /// <param name="context"></param>
         public void RegisterConnectionId(HubLifetimeContext context)
             => ConnectionIds[context.Hub.GetType()] = context.Context.ConnectionId;
     }

```

Reply:
Applied with minor alterations.